File: WebSpider_src/Mf/MfService/Service/WebSpider/WebSpider.cs
using System;
using System.Collections;
using System.Text.RegularExpressions;
using System.IO;
using System.Net;
using System.Windows.Forms;

using Mf.Util;

namespace Mf.Service.WebSpider
{
   /// <summary>
   /// Summary description for Spider.
   /// </summary>
   public class WebSpider
   {
      private        Uri      	      m_startUri;
      private        Uri   		      m_baseUri;
      private        int	            m_uriProcessedCountMax;
      private        int			      m_uriProcessedCount;
      private        bool              m_keepWebContent;

      private        Queue    	      m_webPagesPending;
      private        Hashtable	      m_webPages;
      private        IWebPageProcessor m_webPageProcessor;

      public WebSpider( string startUri ) : 
         this ( startUri, -1 ) { }

      public WebSpider( string startUri, int uriProcessedCountMax ) : 
         this ( startUri, "", uriProcessedCountMax, false, new WebPageProcessor( ) ) { }

      public WebSpider( string startUri, string baseUri, int uriProcessedCountMax ) : 
         this ( startUri, baseUri, uriProcessedCountMax, false, new WebPageProcessor( ) ) { }

      public WebSpider( string startUri, string baseUri, int uriProcessedCountMax, bool keepWebContent, IWebPageProcessor webPageProcessor )
      {
         StartUri 	               = new Uri( startUri );
         
         // In future this could be null and will process cross-site, but for now must exist
         BaseUri  	               = new Uri( Is.EmptyString( baseUri ) ? m_startUri.GetLeftPart( UriPartial.Authority ) : baseUri );

         UriProcessedCountMax       = uriProcessedCountMax;
         KeepWebContent             = keepWebContent;

         m_webPagesPending          = new Queue( );
         m_webPages                 = new Hashtable( );

         m_webPageProcessor         = webPageProcessor;

         m_webPageProcessor.ContentHandler += new WebPageContentDelegate( this.HandleLinks );
      }

      #region public interface
      public void Execute( )
      {
         UriProcessedCount = 0;

         DateTime start = DateTime.Now;

         Console.WriteLine( "======================================================================================================" );
         Console.WriteLine( "Proccess URI: " + m_startUri.AbsoluteUri );
         Console.WriteLine( "Start At    : " + start );
         Console.WriteLine( "------------------------------------------------------------------------------------------------------" );

         AddWebPage( StartUri, StartUri.AbsoluteUri );

         try
         {
            while ( WebPagesPending.Count > 0 && 
               ( UriProcessedCountMax == -1 || UriProcessedCount < UriProcessedCountMax ) )
            {
               Console.WriteLine( "Max URI's: {0}, Processed URI's: {1}, Pending URI's: {2}", UriProcessedCountMax, UriProcessedCount, WebPagesPending.Count );

               WebPageState state = (WebPageState)m_webPagesPending.Dequeue( );

               m_webPageProcessor.Process( state );

               if ( ! KeepWebContent )
               {
                  state.Content = null;
               }

               UriProcessedCount++;
            }
         }
         catch (Exception ex)
         {
            Console.WriteLine( "Failure while running web spider: " + ex.ToString( ) );
         }

         DateTime end = DateTime.Now;
         float elasped = (end.Ticks - start.Ticks)/10000000;

         Console.WriteLine( "------------------------------------------------------------------------------------------------------" );
         Console.WriteLine( "URI Finished   : " + m_startUri.AbsoluteUri );
         Console.WriteLine( "Pages Processed: " + UriProcessedCount );
         Console.WriteLine( "Pages Pending  : " + WebPagesPending.Count );
         Console.WriteLine( "End At         : " + end );
         Console.WriteLine( "Elasped Time   : {0} seconds", elasped );
         Console.WriteLine( "======================================================================================================" );
      }

      public void HandleLinks( WebPageState state )
      {
         if ( state.ProcessInstructions.IndexOf( "Handle Links" ) != -1 )
         {
            int   counter  = 0;
            Match m        = RegExUtil.GetMatchRegEx( RegularExpression.UrlExtractor, state.Content );

            while( m.Success )
            {
               if ( AddWebPage( state.Uri, m.Groups["url"].ToString( ) ) )
               {
                  counter++;
               }

               m = m.NextMatch( );
            }

            Console.WriteLine( "           : {0} new links were added", counter );
         }
      }

      private bool AddWebPage( Uri baseUri, string newUri )
      {
         // Remove any anchors
         string   url      = StrUtil.LeftIndexOf( newUri, "#" );   

         // Construct a Uri, using the current page Uri as a base reference
         Uri      uri      = new Uri( baseUri, url );

         if ( ! ValidPage( uri.LocalPath ) || m_webPages.Contains( uri ) )
         {
            return false;
         }
         WebPageState state = new WebPageState( uri );

         // Only process links for pages within the same site. 
         if ( uri.AbsoluteUri.StartsWith( BaseUri.AbsoluteUri ) )
         {
            state.ProcessInstructions += "Handle Links";
         }

         m_webPagesPending.Enqueue  ( state );
         m_webPages.Add             ( uri, state );
         
         return true;
      }
      #endregion

      #region local interface
      private static readonly string[] m_validPageExtensions = new string[]
            {"html", "php", "asp", "htm", "jsp", "shtml", "php3", "aspx", "pl", "cfm" };

      // A page is considered valid if there is no extension or if the
      // last character is a /.  If there is an extension then the page
      // is considered valid if that extension is classed as valid.
      private bool ValidPage( string path ) 
      { 
         int   pos      = path.IndexOf( "." );

         if ( pos == -1 || path[ path.Length-1 ] == 47 ) /*.ToString( ).Equals( "/" )*/ 
         {
            return true;
         }

         string uriExt = StrUtil.RightOf( path, pos ).ToLower( );

         // Uri ends in an extension
         foreach ( string ext in m_validPageExtensions )
         {
            if ( uriExt.Equals( ext ) )
            {
               return true;
            }
         }

         return false;
      }
      #endregion

      #region properties
      public IWebPageProcessor WebPageProcessor
      {
         get
         {
            return m_webPageProcessor;
         }
         set
         {  
            m_webPageProcessor = value;
         }
      }
      public Uri StartUri
      {
         get
         {
            return m_startUri;
         }
         set
         {  
            m_startUri = value;
         }
      }
      public Uri BaseUri
      {
         get
         {
            return m_baseUri;
         }
         set
         {  
            m_baseUri = value;
         }
      }
      private int UriProcessedCount
      {
         get
         {
            return m_uriProcessedCount;
         }
         set
         {
            m_uriProcessedCount = value;
         }
      }
      public int UriProcessedCountMax
      {
         get
         {
            return m_uriProcessedCountMax;
         }
         set
         {  
            m_uriProcessedCountMax = value;
         }
      }
      public bool KeepWebContent
      {
         get
         {
            return m_keepWebContent;
         }
         set
         {  
            m_keepWebContent = value;
         }
      }
      public Hashtable WebPages
      {
         get
         {
            return m_webPages;
         }
         set
         {
            m_webPages = value;
         }
      }
      private Queue WebPagesPending
      {
         get
         {
            return m_webPagesPending;
         }
         set
         {
            m_webPagesPending = value;
         }
      }
      #endregion
   }
}

File: WebSpider_src/Mf/MfServiceVb/Vb/Service/WebSpider/WebSpider.vb
Imports System
Imports System.Collections
Imports System.Text.RegularExpressions
Imports System.IO
Imports System.Net
Imports System.Windows.Forms

Imports Mf.Util

Namespace Mf.Service.WebSpider

   Public Class WebSpider
      Private m_startUri As Uri
      Private m_baseUri As Uri
      Private m_uriProcessedCountMax As Integer
      Private m_uriProcessedCount As Integer
      Private m_keepWebContent As Boolean

      Private m_webPagesPending As Queue
      Private m_webPages As Hashtable
      Private m_webPageProcessor As IWebPageProcessor


      Public Sub New(ByVal startUri As String)
         MyClass.New(startUri, -1)
      End Sub 'New

      Public Sub New(ByVal startUri As String, ByVal uriProcessedCountMax As Integer)
         MyClass.New(startUri, "", uriProcessedCountMax, False, New WebPageProcessor)
      End Sub 'New

      Public Sub New(ByVal startUri As String, ByVal baseUri As String, ByVal uriProcessedCountMax As Integer)
         MyClass.New(startUri, baseUri, uriProcessedCountMax, False, New WebPageProcessor)
      End Sub 'New

      Public Sub New(ByVal l_startUri As String, ByVal l_baseUri As String, ByVal l_uriProcessedCountMax As Integer, ByVal l_keepWebContent As Boolean, ByVal l_webPageProcessor As IWebPageProcessor)

         StartUri = New Uri(l_startUri)

         ' In future this could be null and will process cross-site, but for now must exist
         If (l_baseUri Is Nothing Or l_baseUri.Trim().Length() = 0) Then
            BaseUri = New Uri(StartUri.GetLeftPart(UriPartial.Authority))
         Else
            BaseUri = New Uri(l_baseUri)
         End If

         UriProcessedCountMax = l_uriProcessedCountMax
         KeepWebContent = l_keepWebContent

         m_webPagesPending = New Queue
         m_webPages = New Hashtable

         m_webPageProcessor = l_webPageProcessor

         WebPageProcessor.HandleContent = AddressOf Me.HandleLinks

      End Sub 'New

      Public Sub Execute()
         UriProcessedCount = 0

         Dim startTime As DateTime = DateTime.Now

         Console.WriteLine("======================================================================================================")
         Console.WriteLine(("Proccess URI: " + m_startUri.AbsoluteUri))
         Console.WriteLine(("Start At    : " + startTime))
         Console.WriteLine("------------------------------------------------------------------------------------------------------")

         AddWebPage(StartUri, StartUri.AbsoluteUri)

         Try
            While WebPagesPending.Count > 0 And (UriProcessedCountMax = -1 Or UriProcessedCount < UriProcessedCountMax)
               Console.WriteLine("Max URI's: {0}, Processed URI's: {1}, Pending URI's: {2}", UriProcessedCountMax, UriProcessedCount, WebPagesPending.Count)

               Dim state As WebPageState = CType(m_webPagesPending.Dequeue(), WebPageState)

               m_webPageProcessor.Process(state)

               If Not KeepWebContent Then
                  state.Content = Nothing
               End If

               UriProcessedCount += 1
            End While
         Catch ex As Exception
            Console.WriteLine(("Failure while running web spider: " + ex.ToString()))
         End Try

         Dim endTime As DateTime = DateTime.Now
         Dim elasped As Single = (endTime.Ticks - startTime.Ticks) / 10000000

         Console.WriteLine("------------------------------------------------------------------------------------------------------")
         Console.WriteLine("URI Finished   : " + m_startUri.AbsoluteUri)
         Console.WriteLine("Pages Processed: " + UriProcessedCount.ToString())
         Console.WriteLine("Pages Pending  : " + WebPagesPending.Count.ToString())
         Console.WriteLine("End At         : " + endTime)
         Console.WriteLine("Elasped Time   : {0} seconds", elasped)
         Console.WriteLine("======================================================================================================")
      End Sub 'Execute


      Public Sub HandleLinks(ByVal state As WebPageState)
         Console.WriteLine("Delegate is called")
         If Not state.ProcessInstructions.IndexOf("Handle Links") = -1 Then
            Console.WriteLine("Delegate is called - Full")
            Dim counter As Integer = 0
            Dim m As Match = RegExUtil.GetMatchRegEx(RegularExpression.UrlExtractor, state.Content)

            While m.Success
               If AddWebPage(state.Uri, m.Groups("url").ToString()) Then
                  counter += 1
               End If

               m = m.NextMatch()
            End While

            Console.WriteLine("           : {0} new links were added", counter)
         End If
      End Sub 'HandleLinks


      Private Function AddWebPage(ByVal l_baseUri As Uri, ByVal newUri As String) As Boolean
         ' Remove any anchors
         Dim url As String = StrUtil.LeftIndexOf(newUri, "#")

         ' Construct a Uri, using the current page Uri as a base reference

         Dim uri As New Uri(l_baseUri, url)

         If Not ValidPage(uri.LocalPath) Or m_webPages.Contains(uri) Then
            Return False
         End If
         Dim state As New WebPageState(uri)

         ' Only do a full process for pages within the same site

         If (uri.AbsoluteUri.StartsWith(BaseUri.AbsoluteUri)) Then
            state.ProcessInstructions += "Handle Links"
         End If

         '         _.P( "URI: {0}, Status: {1}", state.Uri, state.WebPageStatus );
         m_webPagesPending.Enqueue(state)
         m_webPages.Add(uri, state)

         Return True
      End Function 'AddWebPage

      Private Shared m_validPageExtensions() As String = {"html", "php", "asp", "htm", "jsp", "shtml", "php3", "aspx", "pl", "cfm"}

      Private Function ValidPage(ByVal path As String) As Boolean
         Dim pos As Integer = path.IndexOf(".")

         If pos = -1 Or path.Chars(path.Length - 1) = "/" Then   '.ToString( ).Equals( "/" )
            Return True
         End If

         Dim uriExt As String = StrUtil.RightOf(path, pos).ToLower()

         ' Uri ends in an extension
         Dim ext As String
         For Each ext In m_validPageExtensions
            If uriExt.Equals(ext) Then
               Return True
            End If
         Next ext

         Return False
      End Function 'ValidPageExtension

      Public Property WebPageProcessor() As IWebPageProcessor
         Get
            Return m_webPageProcessor
         End Get
         Set(ByVal Value As IWebPageProcessor)
            m_webPageProcessor = Value
         End Set
      End Property

      Public Property StartUri() As Uri
         Get
            Return m_startUri
         End Get
         Set(ByVal Value As Uri)
            m_startUri = Value
         End Set
      End Property

      Public Property BaseUri() As Uri
         Get
            Return m_baseUri
         End Get
         Set(ByVal Value As Uri)
            m_baseUri = Value
         End Set
      End Property

      Private Property UriProcessedCount() As Integer
         Get
            Return m_uriProcessedCount
         End Get
         Set(ByVal Value As Integer)
            m_uriProcessedCount = Value
         End Set
      End Property

      Public Property UriProcessedCountMax() As Integer
         Get
            Return m_uriProcessedCountMax
         End Get
         Set(ByVal Value As Integer)
            m_uriProcessedCountMax = Value
         End Set
      End Property

      Public Property KeepWebContent() As Boolean
         Get
            Return m_keepWebContent
         End Get
         Set(ByVal Value As Boolean)
            m_keepWebContent = Value
         End Set
      End Property

      Public Property WebPages() As Hashtable
         Get
            Return m_webPages
         End Get
         Set(ByVal Value As Hashtable)
            m_webPages = Value
         End Set
      End Property

      Private Property WebPagesPending() As Queue
         Get
            Return m_webPagesPending
         End Get
         Set(ByVal Value As Queue)
            m_webPagesPending = Value
         End Set
      End Property

   End Class 'WebSpider '

End Namespace 'Mf.Service.WebSpider
